File: research/models/logit_jump_diffusion.py
"""Logit Jump-Diffusion for prediction-market prices.

Mirrors src/tools/finance/logit-jump-diffusion.ts.

Reference: *Toward Black–Scholes for Prediction Markets* (arXiv:2510.15205).
Models the price p_t ∈ (0, 1) of a binary prediction-market contract by
working in log-odds space x = logit(p) where the support is unbounded.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Callable, Optional

import numpy as np

Z_MAX = 30.0


def logit(p: float) -> float:
    eps = 1e-12
    clipped = min(1 - eps, max(eps, p))
    return math.log(clipped / (1 - clipped))


def inv_logit(z: float) -> float:
    if z >= Z_MAX:
        return 1 - 1e-12
    if z <= -Z_MAX:
        return 1e-12
    if z >= 0:
        e = math.exp(-z)
        return 1.0 / (1.0 + e)
    e = math.exp(z)
    return e / (1.0 + e)


def ito_martingale_drift(p: float, sigma: float, lam: float, jump_mean: float, jump_std: float) -> float:
    """Itô-Jensen drift correction such that p = sigmoid(x) is a martingale.

    μ_x = -½ σ² (1 - 2p) - λ · ⟨Δp⟩ / (p(1-p))

    Jump compensator integral approximated via Gauss-Hermite 3-point quadrature.
    """
    safe_p = min(1 - 1e-12, max(1e-12, p))
    diffusion_drift = -0.5 * sigma * sigma * (1 - 2 * safe_p)
    if lam <= 0 or (jump_mean == 0 and jump_std == 0):
        return diffusion_drift
    nodes = (-math.sqrt(1.5), 0.0, math.sqrt(1.5))
    weights = (1 / 6, 2 / 3, 1 / 6)
    x = logit(safe_p)
    expected = 0.0
    for w, node in zip(weights, nodes):
        j = jump_mean + math.sqrt(2.0) * jump_std * node
        expected += w * inv_logit(x + j)
    mean_dp = expected - safe_p
    denom = safe_p * (1 - safe_p)
    if denom < 1e-12:
        return diffusion_drift
    return diffusion_drift - (lam * mean_dp) / denom


@dataclass
class LogitJumpDiffusionResult:
    terminal: np.ndarray  # shape (n_paths,)
    paths: Optional[np.ndarray]  # shape (n_paths, days) if requested
    total_jumps: int
    effective_lambda: float


def simulate_logit_jump_diffusion(
    initial_price: float,
    days: int,
    sigma_per_day: float,
    n_paths: int,
    *,
    jump_intensity_per_day: float = 0.0,
    polymarket_jump_prob: Optional[float] = None,
    jump_logit_mean: float = 0.0,
    jump_logit_std: float = 0.0,
    rng: Optional[Callable[[], float]] = None,
    store_paths: bool = False,
) -> LogitJumpDiffusionResult:
    if days < 1:
        raise ValueError("days must be ≥ 1")
    if n_paths < 1:
        raise ValueError("n_paths must be ≥ 1")

    if polymarket_jump_prob is not None:
        lam = max(0.0, min(1.0, polymarket_jump_prob)) / days
    else:
        lam = max(0.0, jump_intensity_per_day)

    if rng is None:
        nprng = np.random.default_rng()
        rng = nprng.random  # type: ignore[assignment]

    initial_logit = logit(initial_price)
    terminal = np.empty(n_paths)
    paths = np.empty((n_paths, days)) if store_paths else None
    total_jumps = 0

    for s in range(n_paths):
        x = initial_logit
        p = inv_logit(x)
        for d in range(days):
            drift = ito_martingale_drift(p, sigma_per_day, lam, jump_logit_mean, jump_logit_std)
            z = _box_muller(rng)
            x += drift + sigma_per_day * z
            if lam > 0 and rng() < lam:
                jz = _box_muller(rng)
                x += jump_logit_mean + jump_logit_std * jz
                total_jumps += 1
            p = inv_logit(x)
            if paths is not None:
                paths[s, d] = p
        terminal[s] = p

    return LogitJumpDiffusionResult(terminal=terminal, paths=paths, total_jumps=total_jumps, effective_lambda=lam)


def _box_muller(rng: Callable[[], float]) -> float:
    u1 = max(rng(), 1e-12)
    u2 = rng()
    return math.sqrt(-2 * math.log(u1)) * math.cos(2 * math.pi * u2)

File: src/tools/finance/logit-jump-diffusion.ts
/**
 * Logit Jump-Diffusion for prediction-market prices.
 *
 * Reference: *Toward Black–Scholes for Prediction Markets* (arXiv:2510.15205,
 * Oct 2025). Models the price p_t ∈ (0, 1) of a binary prediction-market
 * contract by working in log-odds space x = logit(p) where the support is
 * unbounded:
 *
 *   dx = μ(p) dt + σ dW + J · dN
 *
 *   μ(p) = -½ σ² (1 - 2p) - λ · ⟨Δp⟩ / (p(1-p))   ← Itô-Jensen drift
 *
 *   J ~ N(μ_J, σ_J²)   in log-odds space
 *   N is a Poisson process with intensity λ.
 *
 * The drift μ(p) is the **martingale-constrained** value: setting it to this
 * Itô-Jensen correction makes E[p_t | p_{t-1}] ≈ p_{t-1} per step, i.e., p
 * is a (discrete-time) martingale up to O(dt²). Without the correction the
 * naive simulation would drift toward 0.5 because of the concavity of the
 * sigmoid.
 *
 * Polymarket-informed intensity: callers can pass `polymarketJumpProb`
 * (the probability the underlying event resolves within `days`) and the
 * function decomposes it into a per-day λ = polymarketJumpProb / days.
 *
 * Mirrors `research/models/logit_jump_diffusion.py`.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LogitJumpDiffusionParams {
  /** Current Polymarket price in (0, 1). Boundary inputs are clipped. */
  initialPrice: number;
  /** Horizon in days (≥ 1). */
  days: number;
  /** Per-day diffusion volatility in *log-odds* units. */
  sigmaPerDay: number;

  /**
   * Per-day Poisson jump intensity λ.
   *   ── If `polymarketJumpProb` is set, that takes precedence and λ is set
   *      to `polymarketJumpProb / days`.
   *   ── Otherwise this raw intensity is used directly.
   */
  jumpIntensityPerDay?: number;
  /** Polymarket-implied total probability of one jump event over the horizon. */
  polymarketJumpProb?: number;

  /** Mean log-odds shift when a jump fires. Sign encodes direction. */
  jumpLogitMean?: number;
  /** Std-dev of the jump's log-odds shift. */
  jumpLogitStd?: number;

  /** Number of MC paths. */
  nPaths: number;
  /** Optional RNG `() => [0,1)`. Defaults to Math.random for production use. */
  rng?: () => number;
  /** Whether to retain the full (nPaths × days) trajectory matrix. */
  storePaths?: boolean;
}

export interface LogitJumpDiffusionResult {
  /** Terminal price for each path, length = nPaths. */
  terminal: number[];
  /** Per-day price trajectory if `storePaths` was true. */
  paths?: number[][];
  /** Total number of jump events fired across all paths and timesteps. */
  totalJumps: number;
  /** Per-day intensity actually used. */
  effectiveLambda: number;
}

// ---------------------------------------------------------------------------
// Logit / inverse-logit (numerically safe)
// ---------------------------------------------------------------------------

const Z_MAX = 30; // exp(30)/(1+exp(30)) ≈ 1 to ~13 digits — good enough.

export function logit(p: number): number {
  const eps = 1e-12;
  const clipped = Math.min(1 - eps, Math.max(eps, p));
  return Math.log(clipped / (1 - clipped));
}

export function invLogit(z: number): number {
  if (z >= Z_MAX) return 1 - 1e-12;
  if (z <= -Z_MAX) return 1e-12;
  if (z >= 0) {
    const e = Math.exp(-z);
    return 1 / (1 + e);
  }
  const e = Math.exp(z);
  return e / (1 + e);
}

// ---------------------------------------------------------------------------
// Martingale-constrained drift
// ---------------------------------------------------------------------------

/**
 * Itô-Jensen drift correction in log-odds space such that p = sigmoid(x)
 * is a martingale to leading order in dt.
 *
 *   μ_x = -½ σ² (1 - 2p)  +  jump-compensator
 *
 * Jump compensator: -λ · (E_J[sigmoid(x+J) − p]) / (p(1-p)). For a Gaussian
 * jump in logit space we approximate E_J[sigmoid(x+J)] by a Hermite quadrature
 * (3-point) — exact integration is unnecessary; we just need the martingale
 * property to hold to ~O(σ⁴, λ²) which Hermite-3 covers.
 */
export function itoMartingaleDrift(
  p: number,
  sigma: number,
  lambda: number,
  jumpMean: number,
  jumpStd: number,
): number {
  const safeP = Math.min(1 - 1e-12, Math.max(1e-12, p));
  const diffusionDrift = -0.5 * sigma * sigma * (1 - 2 * safeP);
  if (lambda <= 0 || (jumpMean === 0 && jumpStd === 0)) {
    return diffusionDrift;
  }
  // Gauss–Hermite 3-point nodes & weights for ∫ e^{-x²}f(x) dx.
  const nodes = [-Math.sqrt(1.5), 0, Math.sqrt(1.5)];
  const weights = [1 / 6, 2 / 3, 1 / 6];
  const x = logit(safeP);
  let expected = 0;
  for (let i = 0; i < 3; i++) {
    const j = jumpMean + Math.SQRT2 * jumpStd * nodes[i];
    expected += weights[i] * invLogit(x + j);
  }
  const meanDp = expected - safeP;
  const denom = safeP * (1 - safeP);
  if (denom < 1e-12) return diffusionDrift;
  return diffusionDrift - (lambda * meanDp) / denom;
}

// ---------------------------------------------------------------------------
// Simulator
// ---------------------------------------------------------------------------

function boxMuller(rng: () => number): number {
  const u1 = Math.max(rng(), 1e-12);
  const u2 = rng();
  return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
}

export function simulateLogitJumpDiffusion(params: LogitJumpDiffusionParams): LogitJumpDiffusionResult {
  const {
    initialPrice,
    days,
    sigmaPerDay,
    jumpLogitMean = 0,
    jumpLogitStd = 0,
    nPaths,
    rng = Math.random,
    storePaths = false,
  } = params;

  if (!(days >= 1)) throw new Error('simulateLogitJumpDiffusion: days must be ≥ 1');
  if (!(nPaths >= 1)) throw new Error('simulateLogitJumpDiffusion: nPaths must be ≥ 1');

  const lambda = params.polymarketJumpProb !== undefined
    ? Math.max(0, Math.min(1, params.polymarketJumpProb)) / days
    : Math.max(0, params.jumpIntensityPerDay ?? 0);

  const initialLogit = logit(initialPrice);
  const terminal: number[] = new Array(nPaths);
  const paths: number[][] | undefined = storePaths ? new Array(nPaths) : undefined;
  let totalJumps = 0;

  for (let s = 0; s < nPaths; s++) {
    let x = initialLogit;
    let p = invLogit(x);
    const trajectory: number[] | undefined = storePaths ? new Array(days) : undefined;
    for (let d = 0; d < days; d++) {
      const drift = itoMartingaleDrift(p, sigmaPerDay, lambda, jumpLogitMean, jumpLogitStd);
      const z = boxMuller(rng);
      x += drift + sigmaPerDay * z;
      if (lambda > 0 && rng() < lambda) {
        const jz = boxMuller(rng);
        x += jumpLogitMean + jumpLogitStd * jz;
        totalJumps++;
      }
      p = invLogit(x);
      if (trajectory) trajectory[d] = p;
    }
    terminal[s] = p;
    if (paths && trajectory) paths[s] = trajectory;
  }

  return { terminal, paths, totalJumps, effectiveLambda: lambda };
}
